docs/csharp/tutorials/pattern-matching.md

```diff
@@ -0,0 +1,325 @@
+
+The classic object-oriented design would call for creating types in your application that represent each data type from those multiple data sources. Then, your application would work with those new types, build inheritance hierarchies, create virtual methods, and implement abstractions. Those techniques work, and sometimes they are the best tools. Other times you can write less code. You can write more clear code using techniques that separate the data from the operations that manipulate that data.
+
+In this tutorial, you'll create and explore an application that takes incoming data from several external sources for a single scenario. You'll see how **pattern matching** provides an efficient way to consume and process that data in ways that weren't part of the original system.
+
+Consider a major metro area that is using tolls and peak time pricing to manage traffic. You write an application that calculates tolls for a vehicle based on its type. Later enhancements incorporate pricing based on the number of occupants in the vehicle. Further enhancements add pricing based on the time and the day of the week.
+
+From that brief description, you may have quickly sketched out an object hierarchy to model this system. If you're building the entire system, and these were the only requirements, that would be a good solution. That's not the situation. This metro area had already been using software to manage vehicle registrations for years. They bought different systems from different suppliers for personal cars, taxis, buses, and delivery trucks. Different systems all provide classes to represent the data they export. But these different systems don't provide a single object model for you to use. In this tutorial, you'll use these simplified classes for the vehicle classes from these external systems, as shown in the following code:
+
+[!code-csharp[ExternalSystems](../../../samples/csharp/tutorials/patterns/start/toll-calculator/ExternalSystems.cs)]
+
+You can download the starter code from the [dotnet/samples](https://github.com/dotnet/samples/tree/master/csharp/tutorials/patterns/start) GitHub repository. You can see that the vehicle classes are from different systems, and are in different namespaces. No common base class, other than `System.Object` can be leveraged.
+
+## Pattern matching designs
+
+The scenario used in this tutorial highlights the kinds of problems that are well suited to use pattern matching to solve: 
+
+- The objects you need to work with aren't in an object hierarchy that matches your goals. You may be working with classes that are part of unrelated systems.
+- The functionality you're adding isn't part of the core abstraction for these classes. The toll paid by a vehicle *changes* for different types of vehicles, but the toll isn't a core function of the vehicle.
+
+The pattern matching features in C# make it easier to build software when the *shape* of the data and the *operations* on that data are not described together.
```

Comment:
This makes it sound like pattern matching is **only** useful "when the *shape* of the data..." I suspect this is because of the word **when**. Maybe this should be reversed?

```suggestion
When the *shape* of the data and the *operations* on that data are not described together, the pattern matching features in C# make it easier to work with. 
```